services/placement/staged_placement_watcher.go

```diff
@@ -0,0 +1,157 @@
+	placement services.ActiveStagedPlacement
+}
+
+// NewStagedPlacementWatcher creates a new staged placement watcher.
+func NewStagedPlacementWatcher(opts services.StagedPlacementWatcherOptions) services.StagedPlacementWatcher {
+	watcher := &stagedPlacementWatcher{
+		nowFn:         opts.ClockOptions().NowFn(),
+		placementOpts: opts.ActiveStagedPlacementOptions(),
+		proto:         &schema.PlacementSnapshots{},
+	}
+	watcher.doneFn = watcher.onActiveStagedPlacementDone
+
+	valueOpts := runtime.NewOptions().
+		SetInstrumentOptions(opts.InstrumentOptions()).
+		SetInitWatchTimeout(opts.InitWatchTimeout()).
+		SetKVStore(opts.StagedPlacementStore()).
+		SetUnmarshalFn(watcher.toStagedPlacement).
+		SetProcessFn(watcher.process)
+	watcher.Value = runtime.NewValue(opts.StagedPlacementKey(), valueOpts)
+	return watcher
+}
+
+func (t *stagedPlacementWatcher) Watch() error {
+	t.Lock()
+	if t.state != placementWatcherNotWatching {
+		t.Unlock()
+		return errPlacementWatcherIsWatching
+	}
+	t.state = placementWatcherWatching
+	t.Unlock()
+
+	// NB(xichen): we watch the placementWatcher updates outside the lock because
+	// otherwise the initial update will trigger the process() callback,
+	// which attempts to acquire the same lock, causing a deadlock.
+	return t.Value.Watch()
+}
+
+func (t *stagedPlacementWatcher) ActiveStagedPlacement() (services.ActiveStagedPlacement, services.DoneFn, error) {
+	t.RLock()
+	if t.state != placementWatcherWatching {
+		t.RUnlock()
+		return nil, nil, errPlacementWatcherIsNotWatching
+	}
+	return t.placement, t.doneFn, nil
+}
+
+func (t *stagedPlacementWatcher) Unwatch() error {
+	t.Lock()
+	if t.state != placementWatcherWatching {
+		t.Unlock()
+		return errPlacementWatcherIsNotWatching
+	}
+	t.state = placementWatcherNotWatching
+	if t.placement != nil {
+		t.placement.Close()
+	}
+	t.placement = nil
+	t.Unlock()
+
+	// NB(xichen): we unwatch the updates outside the lock to avoid deadlock
+	// due to placementWatcher contending for the runtime value lock and the
+	// runtime updating goroutine attempting to acquire placementWatcher lock.
+	t.Value.Unwatch()
+	return nil
+}
+
+func (t *stagedPlacementWatcher) onActiveStagedPlacementDone() { t.RUnlock() }
+
+func (t *stagedPlacementWatcher) toStagedPlacement(value kv.Value) (interface{}, error) {
+	t.Lock()
+	defer t.Unlock()
+
+	if t.state != placementWatcherWatching {
+		return nil, errPlacementWatcherIsNotWatching
+	}
+	if value == nil {
+		return nil, errNilValue
+	}
+	t.proto.Reset()
+	if err := value.Unmarshal(t.proto); err != nil {
+		return nil, err
+	}
+	version := value.Version()
+	return NewStagedPlacement(version, t.proto, t.placementOpts)
+}
+
+func (t *stagedPlacementWatcher) process(value interface{}) error {
+	t.Lock()
+	defer t.Unlock()
+
+	if t.state != placementWatcherWatching {
+		return errPlacementWatcherIsNotWatching
+	}
+	ps := value.(services.StagedPlacement)
+	placement := ps.ActiveStagedPlacement(t.nowFn().UnixNano())
```

Comment:
ActiveStagedPlacement will return the currently active placement alongside all placements in the future. To get the current active placement, users need to call `.ActivePlacement()` on the active staged placement. This way if the time passes the future cutover time, the correct active placement will be chosen.

There is still an issue with clock skews, but I'll handle that in a separate diff when I start working on topology changes.